File: lec_3/heap.cpp
#include <iostream>
#include <vector>
#include <cctype>
#include <string>
#include <sstream>

using namespace std;


int parent(int i){
	return i/2;
}

int left(int i){
	return 2*(i) + 1;
}

int right(int i){
	return 2*(i) + 2;
}

void printHeap(vector<int> &xheap){
	int i = 0;
	while(i < parent(xheap.size())){
		cout << "node " << xheap[i] << "----" << endl;
		cout << "LEFT " << xheap[left(i)] << " -+-" << endl;
		cout << "RIGHT " << xheap[right(i)] << "//" << endl << endl;
		i++;
	}
}
	


void maxHeapify(vector<int> &a, int i){
	int l = left(i);
	int r = right(i);
	int largest = i;

	if(l < a.size() && a[l] > a[largest])
		largest = l;
	if(r < a.size() && a[r] > a[largest])
		largest = r;

	if(largest != i){
		//swap largest and i
		int temp = a[largest];
		a[largest] = a[i];
		a[i] = temp;
		maxHeapify(a, largest);
	}

}

void buildMaxHeap(vector<int> &a){
	//take the bottom up approach
	for(int i = (a.size() /2 ) +1; i >= 0; --i){
		cout << "testing:: " << a[i] << endl;
		cout << "LEFT " << a[left(i)] << " -+-" << endl;
		cout << "RIGHT " << a[right(i)] << "//" << endl << endl;
		maxHeapify(a, i);
	}
}


int main(void){

	vector<int> xheap;



	int c;
	size_t s = 0;
	while((c = cin.get()) != EOF){
		if(isdigit(c)){
			xheap.push_back(c - '0');
		}
	}

	for(int i = 0; i < s; i++){
		cout << xheap[i] << endl;
	}

	 printHeap(xheap);
	buildMaxHeap(xheap);
	cout << "built" << endl;
	printHeap(xheap);

	// cout << xheap[(xheap.size() / 2 )- 1];
}

File: lec_3/heap.py
import math

def left(i):
	return 2*(i) + 1

def right(i):
	return 2*i + 2

def parent(i):
	return i /2;


# def max_heapify(a, i):
# 	# print("heapifyiing::: ")
# 	# print(a)
# 	l = left(i)
# 	r = right(i)
# 	largest = i

# 	if l < len(a) and a[l] > a[largest]:
# 		largest = l;
# 	if r < len(a) and a[r] > a[largest]:
# 		largest = r;

# 	if largest != i:
# 		temp = a[largest]
# 		a[largest] = a[i]
# 		a[i] = temp
# 		max_heapify(a, largest)


def max_heapify(a, size, i):
	l = left(i)
	r = right(i)
	largest = i
	if l < size and a[l] > a[largest]:
		largest = l
	if r < size and a[r] > a[largest]:
		largest = r

	if largest != i:
		temp = a[largest]
		a[largest] = a[i]
		a[i] = temp
		max_heapify(a, size, largest)



def build_max_heap(a):
	for i in range(int(len(a)/2 -1), -1, -1):
		max_heapify(a, len(a), i)

def extract_max(a):
	max = a[0]
	a[0] = a[len(a)-1]
	a = a[0:len(a)-1]
	max_heapify(a, len(a), 0)
	return max

def heap_delete(a, i):
 	a[i] = a[len(a)-1]
 	a = a[0:len(a)-1]
 	print("a is")
 	print(a)
 	max_heapify(a, len(a),i)
 	print("Aget")
 	print(a)

def heapsort(a):
	build_max_heap(a)
	print("here")
	for i in range(len(a) - 1, -1, -1):
		temp = a[0]
		a[0] = a[i]
		a[i] = temp
		max_heapify(a, i-1, 0)


def main():
	heap = [1,8, 5, 3, 2, 9]
	
	print(heap)
	build_max_heap(heap)
	# heapsort(heap);
	# create a
	heap_delete(heap, 2)
	# print(extract_max(heap))



if __name__ == "__main__":
	main()